src/Tools/PrepareTests/TestDiscovery.cs

```diff
@@ -0,0 +1,128 @@
+
+        var stopwatch = new Stopwatch();
+        stopwatch.Start();
+        vstestConsoleWrapper.DiscoverTests(assemblies, @"<RunSettings><RunConfiguration><MaxCpuCount>0</MaxCpuCount></RunConfiguration></RunSettings>", discoveryHandler);
+        stopwatch.Stop();
+
+        var tests = discoveryHandler.GetTests();
+
+        Console.WriteLine($"Discovered {tests.Length} tests in {stopwatch.Elapsed}");
+
+        stopwatch.Restart();
+        var testGroupedByAssembly = tests.GroupBy(test => test.Source);
+        foreach (var assemblyGroup in testGroupedByAssembly)
+        {
+            var directory = Path.GetDirectoryName(assemblyGroup.Key);
+
+            // Tests with combinatorial data are output multiple times with the same fully qualified test name.
+            // We only need to include it once as run all combinations under the same filter.
+            var testToWrite = assemblyGroup.Select(test => test.FullyQualifiedName).Distinct().ToList();
+
+            using var fileStream = File.Create(Path.Combine(directory!, "testlist.json"));
+            JsonSerializer.Serialize(fileStream, testToWrite);
+        }
+        stopwatch.Stop();
+        Console.WriteLine($"Serialized tests in {stopwatch.Elapsed}");
+    }
+
+    private class DiscoveryHandler : ITestDiscoveryEventsHandler
+    {
+        private readonly ConcurrentBag<TestCase> _tests = new();
+        private bool _isComplete = false;
+
+        public void HandleDiscoveredTests(IEnumerable<TestCase>? discoveredTestCases)
+        {
+            if (discoveredTestCases != null)
+            {
+                foreach (var test in discoveredTestCases)
+                {
+                    _tests.Add(test);
+                }
+            }
+        }
+
+        public void HandleDiscoveryComplete(long totalTests, IEnumerable<TestCase>? lastChunk, bool isAborted)
+        {
+            if (lastChunk != null)
+            {
+                foreach (var test in lastChunk)
+                {
+                    _tests.Add(test);
+                }
+            }
+
+            _isComplete = true;
+        }
+
+        public void HandleLogMessage(TestMessageLevel level, string? message)
+        {
+            Console.WriteLine(message);
+        }
+
+        public void HandleRawMessage(string rawMessage)
+        {
+        }
+
+        public ImmutableArray<TestCase> GetTests()
+        {
+            Contract.Assert(_isComplete);
+            return _tests.ToImmutableArray();
+        }
+    }
+
+
+    private static List<string> GetAssemblies(string binDirectory, bool isUnix)
+    {
+        var unitTestAssemblies = Directory.GetFiles(binDirectory, "*UnitTests.dll", SearchOption.AllDirectories).Where(ShouldInclude);
+        return unitTestAssemblies.ToList();
+
+        bool ShouldInclude(string path)
+        {
+            if (isUnix)
+            {
+                return Path.GetFileName(Path.GetDirectoryName(path)) != "net472";
```

Comment:
I can pass this in from the script / yaml if you'd prefer, but I'll still have to either list the tfm's we want to discover (could change if we start multi-targeting a different tfm for example) or the ones we want to exclude (which as you mention could change).  

Aternatively, if you have a suggestion on some metadata I could look at in the assembly to determine if its compatible with the current platform, that would be preferrable, I wasn't able to immediately come up with a great way.  

I realize now my comment here got lost in some of the changes - but basically the unix build builds multitargeted  projects which ends up building the net472 dlls.  RUnning discovery on them on unix fails.

Reply:
I added a comment here - but didn't end up changing it.  Some notes
1.  I don't want to filter by valid tfm here because those seem more likely to change than invalid ones.  And if they change we'll end up with silent failures of tests not being discovered.
2.  I could move the excluded tfm up to the yaml, but that doesn't necessarily seem better (up to you though).
3.  If the .net framework tfm changes (e.g. we change to net48) then we will get an error during discovery and we can update this, which imho is better than the silent failure.


I could also consider using SRM to check if the TargetFramework attribute contains 'NETFramework' but not sure if that is worth doing.  Happy to hear other thoughts.